sparse/client.go

```diff
@@ -68,70 +88,79 @@ func SyncFile(localPath string, remote string, timeout int, directIO bool) error
 }
 
 func (client *syncClient) syncFileContent(file FileIoProcessor, fileSize int64, directIO bool) error {
-	exts, err := GetFiemapExtents(file)
-	if err != nil {
-		return err
+	if err := client.openServer(directIO); err != nil {
+		return fmt.Errorf("openServer failed, err: %v", err)
 	}
 
-	err = client.openServer(directIO)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	const MB = 1024 * 1024
+	syncStartTime := time.Now()
+	counts := make([]int, 4)
+	durations := make([]time.Duration, 4)
+	out, errc, err := GetFileLayout(ctx, file)
 	if err != nil {
-		return fmt.Errorf("openServer failed, err: %s", err)
+		return fmt.Errorf("failed to retrieve file layout for file %v err: %v", file.Name(), err)
 	}
 
-	var lastIntervalEnd int64
-	var holeInterval Interval
-
-	for _, e := range exts {
-		interval := Interval{int64(e.Logical), int64(e.Logical + e.Length)}
-
-		// report hole
-		if lastIntervalEnd < interval.Begin {
-			holeInterval = Interval{lastIntervalEnd, interval.Begin}
-			err := client.syncHoleInterval(holeInterval)
-			if err != nil {
-				return fmt.Errorf("syncHoleInterval %s failed, err: %s", holeInterval, err)
+	mutex := &sync.Mutex{}
+	processSegment := func(segment FileInterval) error {
+		segmentStartTime := time.Now()
+		if segment.Kind == SparseHole {
+			if err := client.syncHoleInterval(segment.Interval); err != nil {
+				return fmt.Errorf("syncHoleInterval %s failed, err: %s", segment.Interval, err)
+			}
+		} else if segment.Kind == SparseData {
+			if err := client.syncDataInterval(file, segment.Interval); err != nil {
```

Comment:
We may need to split a huge interval into smaller chunks during sending or calculating the checksum:
https://github.com/longhorn/longhorn/issues/1948#issuecomment-791641308

Reply:
The client.syncDataInterval already does chunking of the data interval (32 * 4K)

Reply:
OK. I forgot that part. Then the current optimization is enough now.